### PR TITLE
Derive dynamic libraries explicitly to avoid linker errors for "release" builds

### PR DESCRIPTION
**Pull request by pusukuri**
Derive dynamic libraries explicitly to avoid linker errors for "release" builds. rdar://108561857

### Motivation:

One of the test fixtures in XPM that we've been using since 2019 succeeds in Debug builds but fails in Release builds with linker errors. The fix is to derive dynamic libraries for linking.

### Modifications:

Derive dynamic libraries explicitly for linking.

### Result:

Now release builds are also successful.

### CODE REVIEW

**Comment by pusukuri** [comment]
@swift-ci test

**Comment by pusukuri** [comment]
@swift-ci test windows

**Review by neonichu — APPROVED** [review]

**Comment by pusukuri** [comment]
@swift-ci test macos

File: Tests/BuildTests/BuildPlanTests.swift [hunk 1]
```diff
@@ -6198,4 +6198,73 @@ final class BuildPlanTests: XCTestCase {
...
+        let result = try BuildPlanResult(plan: BuildPlan(
+            buildParameters: mockBuildParameters(shouldLinkStaticSwiftStdlib: true),
+            graph: graph,
+            fileSystem: fs,
+            observabilityScope: observability.topScope
+        ))
+        result.checkProductsCount(3)
+        result.checkTargetsCount(3)
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "FooLogging" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "BarLogging" })
```
Comment (MaxDesiatov) [inline Tests/BuildTests/BuildPlanTests.swift:6268]: Aren't there more effects to observe here? Do you need to check that dependencies between products and targets were established correctly?

Comment (pusukuri) [reply Tests/BuildTests/BuildPlanTests.swift:6268]: Thank you, @MaxDesiatov. I have added your recommendation.

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Pull request merged**